src/parser/raindocument.rs

```diff
@@ -0,0 +1,1746 @@
+                    self.problems.push(Problem {
+                        msg: "imports can only be stated at top level".to_owned(),
+                        position: imp.position,
+                        code: ErrorCode::InvalidImport,
+                    });
+                }
+            }
+        }
+
+        self.namespace = namespace;
+
+        // find any remaining strings and include them as errors
+        exclusive_parse(&document, &WS_PATTERN, 0, false)
+            .iter()
+            .for_each(|v| {
+                self.problems.push(Problem {
+                    msg: "unexpected token".to_owned(),
+                    position: v.1,
+                    code: ErrorCode::UnexpectedToken,
+                });
+            });
+
+        // resolve dependencies and parse expressions
+        self.process_dependencies();
+
+        let binding_am = AuthoringMeta(vec![]);
+
+        // instantiate rainlang for each expression
+        if self.import_depth == 0 {
+            // assign working words for this instance
+            self.resolve_global_deployer();
+
+            let mut has_exp = false;
+            for b in &mut self.bindings {
+                let mut is_exp = false;
+                if let BindingItem::Exp(_) = b.item {
+                    is_exp = true;
+                    has_exp = true;
+                }
+                if is_exp {
+                    b.item = BindingItem::Exp(RainlangDocument::create(
+                        b.content.clone(),
+                        if let Some(am) = &self.authoring_meta {
+                            Some(am)
+                        } else if self.ignore_undefined_words {
+                            None
+                        } else {
+                            Some(&binding_am)
+                        },
+                        Some(&self.namespace),
+                    ));
+
+                    if let BindingItem::Exp(e) = &b.item {
+                        b.problems.extend(e.problems.iter().map(|p| Problem {
+                            msg: p.msg.clone(),
+                            position: [
+                                p.position[0] + b.content_position[0],
+                                p.position[1] + b.content_position[0],
+                            ],
+                            code: p.code,
+                        }));
+                    }
+
+                    self.namespace.insert(
+                        b.name.clone(),
+                        NamespaceItem::Node(NamespaceNode {
+                            hash: String::new(),
+                            import_index: -1,
+                            element: NamespaceNodeElement::Binding(b.clone()),
+                        }),
+                    );
+                }
+            }
+            let len = self.problems.len();
+            if !has_exp && len > 0 {
+                if self.problems[len - 1].msg == "cannot find any set of words (undefined deployer)"
```

Comment:
rather than duplicated strings should these messages be an enum?

Reply:
not probably, since some msgs would differ based on the context, but would generally have the same error code